### PR TITLE
Feature/add shovel resource

### PR DESCRIPTION
**Pull request by justinbaur**
Submitting to add support for the rabbitmq shovel as the underling rabbit hole lib supports it already.

### CODE REVIEW

**Comment by justinbaur** [comment]
I believe the reason for failure is due in part to the shovel plugin not being enabled when the CI engine runs the ACC tests.  I can update the deploy.sh with the following
`sudo rabbitmq-plugins enable rabbitmq_shovel` but not sure if that apply's to the CI.

I was able to validate locally with rabbitmq running in docker without the plugin enabled and receiving the same error when make testacc was ran.

**Comment by justinbaur** [comment]
My mistake, realized now I should have been testing against the same steps locally that travis ci performs.  

I used a separate rabbitmq instance for testing.  I will be testing again locally and update the scripts folder for required changes to get the shovel plugin enabled in the docker image used.

File: rabbitmq/resource_shovel.go [hunk 1]
```diff
@@ -0,0 +1,228 @@
...
+	shovelDefinition := setShovelDefinition(shovelMap).(rabbithole.ShovelDefinition)
+
+	log.Printf("[DEBUG] RabbitMQ: Attempting to declare shovel %s in vhost %s", shovelName, vhost)
+	resp, err := rmqc.DeclareShovel(vhost, shovelName, shovelDefinition)
+	log.Printf("[DEBUG] RabbitMQ: shovel declartion response: %#v", resp)
+	if err != nil {
+		return err
+	}
+
+	d.SetId(shovelName)
```
Comment (cyrilgdn) [inline rabbitmq/resource_shovel.go]: We need to have a more complex ID as we can create 2 shovels with the same name in 2 vhosts, `terraform import` will not be able to find it.

Other resources uses format like `name@vhost`

File: rabbitmq/resource_shovel.go [hunk 2]
```diff
@@ -0,0 +1,228 @@
...
+
+	return ReadShovel(d, meta)
+}
+
+func ReadShovel(d *schema.ResourceData, meta interface{}) error {
+	rmqc := meta.(*rabbithole.Client)
+
+	vhost := d.Get("vhost").(string)
```
Comment (cyrilgdn) [inline rabbitmq/resource_shovel.go]: `vhost` and `name` should be parsed with the new ID syntax I mentioned above, when `ReadShovel` is called by `terraform import` command, state is empty so `d.Get("vhost").(string)` will be empty.

File: rabbitmq/resource_shovel_test.go [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+	return func(s *terraform.State) error {
+		rmqc := testAccProvider.Meta().(*rabbithole.Client)
+		shovelInfos, err := rmqc.ListShovels()
+		if err != nil {
+			return fmt.Errorf("Error retrieving shovels: %s", err)
+		}
+
+		for _, info := range shovelInfos {
+			if info.Name == name {
+				return fmt.Errorf("shovel still exists: %v", info)
```
Comment (cyrilgdn) [inline rabbitmq/resource_shovel_test.go:72]: Probably need to compare the vhost too.

File: website/docs/r/shovel.html.markdown [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+No further attributes are exported.
+
+## Import
+
+Shovels can be imported using the `name`
+E.g.
+
+```
+terraform import rabbitmq_shovel.test test_shovel
+```
```
Comment (cyrilgdn) [inline website/docs/r/shovel.html.markdown]: Need to update with the new syntax containing the vhost.

**Review by cyrilgdn — CHANGES_REQUESTED** [review]
HI @justinbaur ,

Thanks a lot for your work :+1: 

All my comments are linked to same point, you just need to identify a resource with shovelName@vhost as we can create 2 shovels with the same name in 2 different vhosts.



To enable the plugin for your tests and for Travis, you can simply create a file in [`scripts`](https://github.com/terraform-providers/terraform-provider-rabbitmq/tree/master/scripts) directory containing the list of plugins we want to enable:
```
[rabbitmq_management,rabbitmq_shovel_management].
```
(the trailing point is important :smile: )

And mount it in the container thanks to the [`docker-compose.yml` file](https://github.com/terraform-providers/terraform-provider-rabbitmq/blob/master/scripts/docker-compose.yml):
```bash
$ git diff docker-compose.yml
diff --git scripts/docker-compose.yml scripts/docker-compose.yml
index 7504a17..c6d5f28 100644
--- scripts/docker-compose.yml
+++ scripts/docker-compose.yml
@@ -6,5 +6,7 @@ services:
         environment:
             RABBITMQ_DEFAULT_USER: ${RABBITMQ_USERNAME:-guest}
             RABBITMQ_DEFAULT_PASS: ${RABBITMQ_PASSWORD:-guest}
+        volumes:
+            - ./enabled_plugins:/etc/rabbitmq/enabled_plugins
         ports:
             - 15672:15672

File: rabbitmq/resource_shovel.go (hunk 1 above)
Comment (justinbaur) [reply rabbitmq/resource_shovel.go]: Good call, forgot that they are stored at a global level, I made the fix

File: website/docs/r/shovel.html.markdown (hunk 4 above)
Comment (justinbaur) [reply website/docs/r/shovel.html.markdown]: changed to name@vhost for the import example

File: rabbitmq/resource_shovel.go (hunk 2 above)
Comment (justinbaur) [reply rabbitmq/resource_shovel.go]: Using both name and vhost for ID now

**Comment by justinbaur** [comment]
Not sure why Rabbit 3.6 keeps failing on declaring the queue and exchange in travisci.  I've ran this against the docker-compose with 3.6 on Linux and was able to pass.

File: go.sum [hunk 5]
```diff
@@ -290,9 +290,11 @@ github.com/oklog/run v1.0.0/go.mod h1:dlhp/R75TPv97u0XWUtDeV/lRKWPKSdTuV0TZvrmrQ
 github.com/onsi/ginkgo v1.6.0/go.mod h1:lLunBs/Ym6LB5Z9jYTR76FiuTmxDTDusOGeTQH+WWjE=
 github.com/onsi/ginkgo v1.7.0 h1:WSHQ+IS43OoUrWtD1/bbclrwK8TTH5hzp+umCiuxHgs=
 github.com/onsi/ginkgo v1.7.0/go.mod h1:lLunBs/Ym6LB5Z9jYTR76FiuTmxDTDusOGeTQH+WWjE=
+github.com/onsi/ginkgo v1.8.0 h1:VkHVNpR4iVnU8XQR6DBm8BqYjN7CRzw+xKUbVVbbW9w=
```
Comment (cyrilgdn) [inline go.sum]: Do you know why does this file change? You didn't change any dependencies so it should not :thinking:

Comment (justinbaur) [reply go.sum]: That was a mistake, it was most likely updated via my go plugin in VS Code.  I'll back it out.

Comment (justinbaur) [reply go.sum]: This file would have changed when I merged the changes from "Switch to rabit-hole 2.0.0" commit https://github.com/terraform-providers/terraform-provider-rabbitmq/commit/5470b5ca8630e95a45b84660a861106c26843d34

Would it make more sense to close the PR and rebase on my master and do a new PR to this master?

**Comment by akurz** [comment]
Hi all, 

I'd really be interested in getting this feature pull request in. Is there anything I could help to make this happen @cyrilgdn, @justinbaur?

Cheers,
Andreas

**Comment by justinbaur** [comment]
@akurz I'll be able to fix the conflicts but it still has a change requested review.  @cyrilgdn does anything else need to be changed in this PR?

**Comment by cyrilgdn** [comment]
> does anything else need to be changed in this PR?

@justinbaur We just need me to make the review :facepalm: I have dedicated time for providers next Friday and I'll start with that if I didn't get the time before.

>  Is there anything I could help to make this happen @cyrilgdn

@akurz You can test the PR with your use case if you know how to do it. It would be very helpful.

**Comment by akurz** [comment]
> > does anything else need to be changed in this PR?
> 
> @justinbaur We just need me to make the review  I have dedicated time for providers next Friday and I'll start with that if I didn't get the time before.
> 
> > Is there anything I could help to make this happen @cyrilgdn
> 
> @akurz You can test the PR with your use case if you know how to do it. It would be very helpful.

Sure, I can give it a try @cyrilgdn and will let you know if all works fine for me

File: website/rabbitmq.erb [hunk 6]
```diff
@@ -37,6 +37,9 @@
             <li<%= sidebar_current("docs-rabbitmq-resource-vhost") %>>
               <a href="/docs/providers/rabbitmq/r/vhost.html">rabbitmq_vhost</a>
             </li>
+            <li<%= sidebar_current("docs-rabbitmq-resource-shovel") %>>
+              <a href="/docs/providers/rabbitmq/r/shovel.html">rabbitmq_shovel</a>
+            </li>
```
Comment (cyrilgdn) [inline website/rabbitmq.erb]:
```suggestion
            </li>
```

**Review by cyrilgdn — CHANGES_REQUESTED** [review]
LGTM :+1: 

Just need to rebase on `master`, rabbithole has been upgraded to v2 (so you'll have to modify the `import`). 
You should not have modifications in `go.sum` anymore after the rebase

**Comment by akurz** [comment]
@cyrilgdn,@justinbaur I have successfully tested this pull request - rebased on current master. I was able to create a working shovel setup in one of our test clusters. Unfortunately rabbithole does not seem to support shovel connections with AMQP1.0 links as needed for Azure Eventhub - but that's another story :-)

**Review by cyrilgdn — APPROVED** [review]
@akurz Thanks for your help :+1: 

@justinbaur I allowed myself to rebase & squash this PR for you so we can merge it and include it in the coming release.

**Comment by akurz** [comment]
You're welcome @cyrilgdn! Do you have a rough time line for the next release? I'd also like to send a pull-request that adds the missing Shovel parameters to the resource.

**Comment by cyrilgdn** [comment]
@akurz Actually [v1.4.0](https://github.com/terraform-providers/terraform-provider-rabbitmq/blob/master/CHANGELOG.md#140-july-17-2020) has just been released with the shovel resource :sweat_smile: 

But feel free to open a PR, we can easily trigger new releases.

**Comment by justinbaur** [comment]
@cyrilgdn Perfect! Thanks for doing that. 

@akurz  Thanks for testing as well.

**Pull request merged**